metadata-io/src/main/java/com/linkedin/metadata/entity/DeleteEntityService.java

```diff
@@ -0,0 +1,254 @@
+   * @param relatedUrn        The urn of the related entity in which we want to find the aspect that has a relationship
+   *                          to `urn`.
+   * @param relationType      The relationship type that details how the `urn` entity is related to `relatedUrn` entity.
+   * @param relatedEntitySpec The entity spec of the related entity.
+   * @return An {@link java.util.Optional} object containing the aspect content & respective spec that contains the
+   * relationship between `urn` & `relatedUrn`.
+   */
+  public java.util.Optional<Pair<EnvelopedAspect, AspectSpec>> getAspectReferringTo(final Urn urn, final Urn relatedUrn,
+      final String relationType, final @Nonnull EntitySpec relatedEntitySpec) {
+
+    // Find which aspects are the candidates for the relationship we are looking for
+    final Map<String, AspectSpec> aspectSpecs = relatedEntitySpec.getAspectSpecMap()
+        .entrySet()
+        .stream()
+        .filter(aspectSpec -> !aspectSpec.getValue().findRelationshipFor(relationType, urn.getEntityType()).isEmpty())
+        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    // FIXME: Can we not depend on entity service?
+    final EntityResponse entityResponse;
+    try {
+      entityResponse = _entityService.getEntityV2(relatedUrn.getEntityType(),
+          relatedUrn,
+          aspectSpecs.keySet());
+    } catch (URISyntaxException e) {
+      log.error("Unable to retrieve entity data for relatedUrn " + relatedUrn, e);
+      return java.util.Optional.empty();
+    }
+
+    // Find aspect which contains the relationship with the value we are looking for
+    return entityResponse
+        .getAspects().values()
+        .stream()
+        // Get aspects which contain the relationship field specs found above
+        .filter(aspect -> aspectWithRelationshipTo(urn.getEntityType(), relationType, aspectSpecs, aspect))
+        .map(aspect -> Pair.of(aspect, aspectSpecs.get(aspect.getName())))
+        .findFirst();
+  }
+
+  /**
+   * Utility method that determines whether a given aspect has a relationship of type relationType to a given entity type.
+   * @param entityType    The name of the entity the method checks against.
+   * @param relationType  The name of the relationship to search for.
+   * @param aspectSpecs   A dictionary of relationship types to aspect specs in which to search for a concrete
+   *                      relationship with name=relationType and that targets the entityType passed by parameter.
+   * @param aspect        The aspect in which to search for the relationship.
+   * @return {@code True} if the aspect has a relationship with the intended conditions, {@code False} otherwise.
+   */
+  private boolean aspectWithRelationshipTo(final String entityType, final String relationType,
+      final Map<String, AspectSpec> aspectSpecs, final EnvelopedAspect aspect) {
+
+    final AspectSpec aspectSpec = aspectSpecs.get(aspect.getName());
```

Comment:
This is a bit confusing. 

We should already know the aspectSpec we are expecting to work with when we call this method.. Should not need this if(aspectSpec == null) return false this?

Reply:
I don't like passing around this aspectSpecs map everywhere, in other words